### ISSUE TITLE
PyQt5.QtQml Hook should import QML Libraries

### ISSUE THREAD

**Issue by GLaDOS1105** [pyinstaller/pyinstaller#3711]
`hook-PyQt5.QtQml` should import QML libraries like `hook-PyQt5.QtQuick` because we could only import `PyQt5.QtQml.QQmlApplicationEngine` or `PyQt5.QtQml.QQmlEngine` in Python script to load QML files and create QML application. `QQmlApplicationEngine` is more powerful and more suitable for QtQuick 2.x. For instance,

``` python
# main.py
import sys
from PyQt5.QtCore import QUrl, QObject
from PyQt5.QtGui import QGuiApplication
from PyQt5.QtQml import QQmlApplicationEngine

if __name__ == '__main__':
    app = QGuiApplication(sys.argv)
    engine = QQmlApplicationEngine()
    engine.load(QUrl('some_qml_application_window.qml'))
    if not engine.rootObjects():
        sys.exit(-1)
    res = app.exec_()
    del engine
    sys.exit(res)
```
Thus, `hook-PyQt5.QtQml.py` should have the same contents with `hook-PyQt5.QtQuick.py` in `Pyinstaller/hooks/` or simply add `PyQt5.QtQuick` to the `hiddenimports` in `hook-PyQt5.QtQml.py`. For example,

``` python
#hook-PyQt5.QtQml.py
from PyInstaller.utils.hooks import add_qt5_dependencies

hiddenimports, binaries, datas = add_qt5_dependencies(__file__)
hiddenimports.append('PyQt5.QtQuick')
```

**Comment by bjones1** [comment]
Thanks -- this is very helpful. I'll work on using your sample code to add a test to the test suite, then fix the resulting failure.

**Comment by GLaDOS1105** [comment]
Furthermore, I think both [`Pyinstaller/hooks/hook-PyQt5.QtQuick.py`](https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/hooks/hook-PyQt5.QtQuick.py) and [`Pyinstaller/hooks/hook-PyQt5.QtQml.py`](https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/hooks/hook-PyQt5.QtQml.py) should load all folders in `qmldir` instead of only the directories in the list `dirs` before we could parse the import dependencies in QML files because some QML libraries are interdependent. For example, if we want to use Imagine style, one of the default QtQuick2 theme, to the QML application, it would require `QtGraphicalEffects` libraries in `qmldir` to complete the job. There are only 2 methods to set the default QtQuick2 theme for an application:

1. [Command line argument](https://doc.qt.io/qt-5.11/qtquickcontrols2-styles.html#command-line-argument)
2. [Environment variable](https://doc.qt.io/qt-5.11/qtquickcontrols2-styles.html#environment-variable)

Note that `QQuickStyle::setStyle("Material");` method is unavailable in PyQt5 as PyQt5 does not provide QQuickStyle class in Python. Thus, it could be difficult for Pyinstaller to parse any files to check which theme would be used and what QML libraries is required. The following is the example QML code setting the Imagine style to cause the problem. (Python script is the same as the previous one.)

``` qml
// some_qml_application_window.qml
import QtQuick 2.11
import QtQuick.Controls 2.4
ApplicationWindow {
    visible: true
    ProgressBar {text: "Hello!"}
}
```

After packaging the application by Pyinstaller, use the following command to set the QtQuick2 style and run the executable in `dist/main/`.

``` bash
./main --style imagine
```

It would produce the following error message:

``` bash
QQmlApplicationEngine failed to load component
file:///home/username/pyinstaller/main/dist/main/some_qml_application_window.qml:6 Type ProgressBar unavailable
file:///home/username/pyinstaller/main/dist/main/PyQt5/Qt/qml/QtQuick/Controls.2/Imagine/ProgressBar.qml:41 module "QtGraphicalEffects" is not installed
```

It turns out that some QML libraries in `QtQuick` require other libraries in the different folder (`QtGraphicalEffects`).

Besides, if we load all folders in `qmldir` in the hook, we could also guarantee that `QtCharts` libraries installed by [PyQtChart](https://pypi.org/project/PyQtChart/) (instead of PyQt5 itself) in `qmldir` would be available when the application import QtCharts in QML files.

Indeed, packaging all QML libraries would increase the total size of the distribution. Yet, before we could parse QML files to locate the QML dependencies, loading all QML libraries to the package could make sure there will be no missing QML types.

**Comment by bjones1** [comment]
This also makes a lot of sense. I'll include it in the updates. Thanks!

**Comment by bjones1** [comment]
Would you take a look at how I've implemented this at https://github.com/bjones1/pyinstaller/tree/issue_3711 and verify this works and that the tests make sense?

**Comment by GLaDOS1105** [comment]
That works flawlessly! Thank you for the fix. By the way, I believe that after PR #3693 has been merged, `hook-PySide2.QtQml.py` might also need to be modified in order to load all QML libraries.

**Comment by bjones1** [comment]
Thanks for testing! I agree that #3693 would need a similar fix. Please provide a pull request! (I only focus on maintaining PyQt5, not PySide2.)

**Comment by bjones1** [comment]
Fixed in 46ee8a0605eeb05d0c432201a412630830a71073.
